core/src/test/scala/org/apache/spark/ShuffleSuite.scala

```diff
@@ -383,13 +383,19 @@ abstract class ShuffleSuite extends SparkFunSuite with Matchers with LocalSparkC
     // simultaneously, and everything is still OK
 
     def writeAndClose(
-      writer: ShuffleWriter[Int, Int])(
+      writer: ShuffleWriter[Int, Int],
+      taskContext: TaskContext)(
       iter: Iterator[(Int, Int)]): Option[MapStatus] = {
-      val files = writer.write(iter)
-      writer.stop(true)
+      TaskContext.setTaskContext(taskContext)
```

Comment:
Because there are TaskContext.get() calls in the implementation — it is necessary that we set these on the tests to avoid NPE from not being set.

Reply:
ok, but is there any particular reason the implementations are doing that, rather than using the passed in TaskContext?

in fact its a little confusing because it only pulls the shuffle write metrics, but those are passed directly to `BypassMergeSortShuffleWriter`, so it then looks like `BypassMergeSortShuffleWriter` is forgetting to pass those metrics on to the `ShuffleMapOutputWriter`.  I know we use taskcontext a lot, but because its global state its harder to keep track of, and to know to add to tests (like here).  If it were an argument to the methods / constructors it would be obvious.

anyway I know we already have to call `TaskContext.set()` in a ton of tests as it is assumed to be there, so this isn't a hard requirement, but seems like we can keep it a little more clean.

Reply:
Because the shuffle writer API is itself not given a `TaskContext` object. Or in other words, we didn't want to put `TaskContext` in any of the function signatures in the plugin tree.

Reply:
any particular reason why not?  like this, its an implicit part of the api.  (Its already a public class.)

is the only reason its needed ShuffleWriteMetrics / ShuffleReadMetrics?  then maybe those should become an explicit part of the api?  otherwise the api is even more confusing -- implementors just have to know they should be grabbing the taskcontext and incrementing those metrics.

Reply:
`TaskContext` should only be used for internal implementations, such as the local write implementation, or suppose if Spark provided libraries for async-write implementations. Basically, I'd only expect the `TaskContext` to be used by spark-provided implementations. Otherwise, metrics should be tracked by the implementation's specific metrics systems.

Reply:
But this decision was also made before when we thought the plugin system would be marked as external APIs instead of developer APIs. So I'm more loose on this matter as a whole.

Reply:
I'd imagine that all spark users, regardless of what shuffle implementation their admins have setup, would want to have tasks report shuffle read & write stats.

Reply:
But will they want to use the task context's metrics system? I'd imagine they would want to use whatever custom metrics system that is appropriate for their environment.

Reply:
Basically I would prefer to minimize the API surface area as much as possible. If we can consider `TaskContext.get()` as an implementation detail and that the `TaskContext` isn't absolutely necessary for the API, I'd rather not include it and only add it to the API in a follow-up. Especially since this discussion is primarily centered around tests, which we have seen there are other places in tests where `TaskContext.set` is used anyways 0- so we're not going against existing precedent.

Reply:
though I mentioned this comment on the test, the whole reason was just because it made me worry about the general api design.

Another shuffle implementation may have its own metric system, to monitor that system -- but nonetheless, the actual end user of spark is going to want to see metrics in the Spark UI about the shuffle.  We don't have a way for the alternative shuffle implementation to plugin their own metrics to the UI (nor do I think we want to).  I guess the most important metrics, the number of records & bytes, are recorded outside of the plugin -- but the plugin should be updating the write time metric, regardless of what storage its using.

Reply:
Why can't the higher level writer update the write time?

Regardless, we can add the task context to the API - but we're pushing the number of parameters I think we would like in `createMapOutputWriter` - 5 parameters is quite a large number. Maybe we can group some of the parameters into a structure.